### ISSUE TITLE
Some pictures won't print in BAK and GLO 

### ISSUE THREAD

**Issue by davidc86** [sillsdev/ptx2pdf#732]
The first 3 maps in the SLU BAK book, and the two illustrations in GLO won't print. 
(They never have printed properly in these books. I don't know why. They work in other places.)
I've no clue.

Here is the archive:
[SLU-SelaruNewPublicationTEST-BACK-MATTER-PTXprintArchive.zip](https://github.com/sillsdev/ptx2pdf/files/9602049/SLU-SelaruNewPublicationTEST-BACK-MATTER-PTXprintArchive.zip)

**Comment by davidg-sil** [comment]
Sadly, the archive only seems to include the figures used in the publication, so I find  5 figures in there, all in use in the PDF. (Paul-1... Paul-4 and PaulLokasi5T).
But... looking at the file names in your piclists, I see that some of your images contain brackets, and others spaces. Since I've always been of the opinion those things don't belong in file names (they both make things hard from the command line), my test files would never have included them.

Maybe that's the problem? I'll have a look from my end, too.

**Comment by davidg-sil** [comment]
Quick update.... The TeX macros have no problems with spaces or brackets in file names from a piclist.

**Comment by davidc86** [comment]
The missing pictures are all used in other places in the scripture text and the print there just fine.
The relevant lines from BAK are: 

\m \fig Nuske Ktem ne heitlulswo kyoat Yunus imorifke|alt="The world long ago when Jonah was living" src="PetaYunus_Selaru.tif" size="col" ref="Yun. 1–4"\fig* 

\m \fig Hulasokwe Seike ti kyoat Yesuske imin ti lasmyer khaha ne|alt="The Temple of God" src="LB00250B.tif" size="col" ref="Luk. 4:9"\fig* 

\m \fig Nus o hnu ti lasmyer Palestinake|alt="The cities and towns in the land of Palestina" src="Holy_Land_BW-Selaru.tif" size="col"\fig* 

(None of these filenames have brackets or other odd punctuation, in fact they are the most behaved filenames I use!)

These are the missing picture lines from the GLO book: 

\pc \fig Hulasokwe Sei a raja a Herodes isadirike|src="LB00250B.TIF" size="col" ref="1:0"\fig* 

\pc \fig Lasyery Nelnyelkye|alt="Tabernacle" src="LB00259B.TIF" size="col" ref="1:9aa"\fig* 

(Again, nothing weird about these either.) 

So, any other ideas?

**Comment by davidc86** [comment]
Woohoo! A breakthrough. I carefully selected just the BAK book only, then I "regenerated" the PicList. (Spooky) And this appended the pictures that PTXprint wasn't seeing before. (I ended up with duplicates, but that is way better than PTXprint not seeing the pictures at all.) I deleted the duplicates, so now PTXprint sees the rest of the pictures. Yay!

REQUEST: From this I'm guessing PTXprint does NOT inform the user, "Hey! These SFM files have \fig pictures lines that aren't listed in your PicList file! Are you cool with that?"  This would be a very helpful notification. And even more helpful if it suggested a safe solution. (Like "Do you want me to add the missing \fig lines to this list of pictures?")

I've been stumped by this for months, but I didn't want to bother you guys with it until more important "scripture text formatting" issues others were having had been solved.

Okay, so all the pictures are now showing in PTXprint. 
But sadly, when I print the PDF, only the first picture shows! All the other pictures in BAK are not sent to the PDF.

The WHOLE BAK book is just one page in the final PDF: 

![image](https://user-images.githubusercontent.com/65363206/191518867-dd2b0670-386e-4c17-ab8b-2b055e03e9d5.png)

And the map is printed ABOVE the title! And nothing else comes through. 

The final SFM BAK file looks like this: 

\id BAK ‑ Selaru
\h Gambar o Peta
\toc3 Peta

\toc1 Gambar o Peta-nare
\mt1 Gambar o Peta
\mt2 ti Kitab Lanke
\p Ti ne tatos gambar o peta-nare ma klyosu o nam rakesy ti Kitab Lanke ne.
\m \zfiga|p001\*
\b
\m \zfiga|p002\*
\b
\m \zfiga|p003\*
\b
\m \zfiga|p004\*
\b
\m \zfiga|p005\*
\b
\m \zfiga|p006\*
\b
\m \zfiga|p007\*
\b
\m \zfiga|p008\*
\b

This seems reasonable to me, but maybe PTXprint needs REAL paragraphs to hook the figures to?

**Comment by davidc86** [comment]
SOLVED! I changed the Vertical Position from "TOP" (which was set by the Regenerate process) to "After Paragraph", and now they all appear just fine!!!

![image](https://user-images.githubusercontent.com/65363206/191520217-aac4dc7b-a936-4c26-9f13-8ce68f1fcae1.png)

Woohoo.  So, I guess this can be closed. But I would like to submit a request that PTXprint notifies the user if there is any \fig line in the original SFM files that are not represented in the PicList. Because for on-going projects, figures are being added (and removed) fairly regularly.

**Comment by davidg-sil** [comment]
Great news that you managed to fix it.  In the publications I've been involved in so-far, figures were always added in a separate step to translating, so I've only ever used a separate piclist.

The whole updating/updated pictures/captions in the SFM thing certainly needs figuring out. PTXprint adding figures to the USFM would probably be bad, but someone might expect that correcting the spelling of the caption in PTXprint would change it in Paratext. Someone else might expect the opposite direction to work, and a 3rd person might expect/want them to stay entirely separate....

**Comment by davidc86** [comment]
Well, bummer. I looks like some of the pictures in BAK suddenly stopped showing after I upgraded to v2.2.24.
Here is the Picture page now: Note that the BAK p001, BAK p003, and BAK p004 are not listed anymore.

![image](https://user-images.githubusercontent.com/65363206/191553024-821a4905-a8c2-4f1d-b28b-3061633d2a69.png)

I tried to create an Archive and got this: 

![image](https://user-images.githubusercontent.com/65363206/191553735-41408c76-cea5-4feb-826d-72d8c31888a1.png)

**Comment by davidc86** [comment]
But even with the error it creates an Archive. So I have an archive where the BAK printed correctly and one where it didn't. There was no change in the BAK file itself. Here is a folder comparison (older archive on the left, new one on the right): 

![image](https://user-images.githubusercontent.com/65363206/191578372-4bcded57-5db8-475c-96e9-cf0ba50398e7.png)

And here is a comparison of the two FINAL BAK SFMs (old on the left, new on the right): 

![image](https://user-images.githubusercontent.com/65363206/191578633-8e8d13a9-1db3-4c71-b995-17e18c675858.png)

So the new one deletes out the picture lines.
This is confusing.

**Comment by davidc86** [comment]
Comparing the two PicList files for this configuration we can see where I fixed the problem with the GLO book (got those pictures to show up), but somehow some of the pictures in the BAK book were deleted out of the list.

![image](https://user-images.githubusercontent.com/65363206/191579344-2f865f89-899a-432c-99e8-7edaf100581d.png)

**Comment by davidc86** [comment]
I'm not seeing where this information is maintained. So I had to regenerate the images for BAK and delete duplicates and reset the positions on all of these. 
It would be nice if rather than the options to 1. Append in X book or 2. Add from ALL books (yikes!) how about "Find missing illustrations and fix the PicList" as an option. (Or is the goal to actually never need to do this?)

**Comment by davidc86** [comment]
Here is what I think would help:
1.) Bug:  PTXprint is deleting pictures from the piclist. History: What started this whole issue was I couldn't get pictures to display in the BAK book. Using Regenerate, I finally got all the BAK pictures in, and it printed fine, but then after I generated the piclist for the GLO pictures (or maybe after I updated the version of PTXprint), suddenly 3 of the pictures in BAK were removed from the piclist and no longer printed! This behavior is a bug, and it is what forced me to see what was going on. So even if no changes are made to the Pictures UI, whatever is causing this weirdness needs squashing.
2.) Once Bug #1 is fixed, the user still has no way of knowing that what is in Paratext is actually being reflected in the huge list of pictures. (It was obvious in my BAK book, but sometimes it won't be, like when only one picture is not present, especially when running PTXprint for other folk.) So, like you suggested, a report after the PDF is made that tells the user if there are any mismatches, and what to do about it, would be very helpful.
3.) Bug: Regenerating should not DUPLICATE the pictures in the piclist (finding 6 \fig...\fig* lines produces 12 entries).
4.) Enhancement: Regenerating should be able to ADD NEW pictures without touching existing pictures. I.e. if the \fig line refers to a particular illustration in a particular chapter, which is already in the piclist (same filename, same chapter), it should be skipped. Only new, unknown filenames should be added to the piclist with an "ADD" feature. (NOTE: some pictures are used more than once in a publication, like once in the scripture text and once again in a Glossary or Maps section.) I know there is an Add button, but it just opens the dialog with empty fields. If we could point to or paste a \fig line from the SFM, that would make it easier and protect against typos, etc.
5.) Enhancement: It would be nice to specify a different default location and position for imported pictures (for some books I use "After paragraph" "Centered" exclusively).  

So those are the suggestions and bugs I've come up with.
